---
title: "wSIR workshop"
author: Max Woollard^[mwoo5086@uni.sydney.edu.au]
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{wSIR workshop}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{=html}
<style>
.question {
  padding: 1em;
  background: lightcyan;
  color: black;
  border-radius: 10px;
}
</style>
```

# wSIR: Weighted Sliced Inverse Regression for supervised dimension reduction of spatial transcriptomics and single cell gene expression data

**Authors**

P = presenting, C = contributing to work

Max Woollard (P)$^{1,2,3}$, Pratibha Panwar (C)$^{1,2,3}$, Linh Nghiem (C)$^{1,2}$, Shila Ghazanfar (C)$^{1,2,3}$.

$^1$ Sydney Precision Data Science Centre, University of Sydney,
Australia\
$^2$ School of Mathematics and Statistics, University of Sydney,
Australia
$^3$ Charles Perkins Centre, University of Sydney,
Australia\

<br/> Contact: mwoo5086\@uni.sydney.edu.au

## Overview

Single-cell and spatial gene expression data are very high-dimensional, which makes dimension reduction a crucial step in the analysis pipeline. However, many popular examples of dimension reduction methods, such as Principal Component Analysis (PCA), ignore the purpose of a dimension reduction procedure, which is to assist in a specific downstream analysis. wSIR is a supervised dimension reduction method, which makes use of cells' spatial locations to compute a spatially-informed low-dimensional embedding.

## Description

In this workshop, we will introduce the wSIR method for supervised dimension reduction of spatial transcriptomics and single-cell gene expression data. We will explain the method, and go through some example analyses to demonstrate how the method is applied. We will also perform some simple analyses to demonstrate its utility. 

### Pre-requisites

It is expected that students will have:

- basic knowledge of R syntax, and
- familiarity with single-cell and spatial transcriptomics data.

### Time outline

The expected timing of the workshop:

| Activity                                      | Time |
|-----------------------------------------------|------|
| wSIR method overview                          | 20m  |
| wSIR application: interpretability            | 15m  |
| wSIR application: improved modelling accuracy | 15m  |

### Learning objectives

- Understand the wSIR method.
- Use wSIR to project new single-cell gene expression data into a spatially-informed low-dimensional space.
- Examine the wSIR loadings to derive biological insight
- Use wSIR to improve accuracy in downstream analysis

## Workshop

## Setup

### Installation

```{r}
#install.packages(c("ggplot2", "vctrs", "magrittr", "devtools"))
#devtools::install_github("SydneyBioX/wSIR")
```

### Load packages

```{r}
library(BioCAsia2024wSIR)  # use the same name, no underscores, as in DESCRIPTION
library(ggplot2)
library(vctrs)
library(wSIR)
library(magrittr)
library(dplyr) # for arrange
```

### Acquire data

We will use spatial transcriptomics data for mouse embryos from https://www.nature.com/articles/s41587-021-01006-2 . We will examine how we can apply the wSIR functions to study this data. This dataset will illustrate how you can apply the package functions to your own data. 

```{r}
#data(embryos_data_red)  # you don't have a data folder
load(system.file("extdata", "embryos_data_red.RData", package="BioCAsia2024wSIR"))

## files this downloads:
# exprs1
# coords1
# exprs2
# coords2
# exprs3
# coords3
```

```{r}
# Ensure coordinate column names are correct
colnames(coords1) <- c("x", "y")
colnames(coords2) <- c("x", "y")
colnames(coords3) <- c("x", "y")

exprs1 <- as.matrix(exprs1)
exprs2 <- as.matrix(exprs2)
exprs3 <- as.matrix(exprs3)
```

```{r}
n1 = nrow(exprs1)
n2 = nrow(exprs2)
n3 = nrow(exprs3)
c(n1, n2, n3)
```

## wSIR method

wSIR is an extension of the supervised dimension reduction technique of Sliced 
Inverse Regression (SIR). 

SIR is an existing supervised dimension reduction method which works by 
grouping together the observations with similar values for the response. For 
spatial transcriptomics data, this means grouping all the cells into a certain 
number of tiles based on their spatial position. For example, if we use 4 
tiles, then the cells in the top right quadrant of the image/tissue go to one 
group, those in the top left to another, and so on. Each of those groups is 
then summarised by averaging the expression level of all cells in each group 
for each of the genes. From there, eigendecomposition is performed on the 
resulting matrix of tile-gene means, then returning the SIR directions and 
SIR scores.

The motivation behind wSIR is that SIR only uses each cell's spatial position 
when we are separating the cells into the given number of groups/tiles. Once 
those groups are created, we lose the fact that some groups may be more 
spatially related (if they come from adjacent tiles) than other groups (if 
they come from opposite sides of the tissue). wSIR uses a weight matrix to 
incorporate the spatial correlation between all pairs of cells in the SIR 
algorithm. This matrix has dimension H*H, where H is the number of tiles, 
and the (i,j)th entry represents the distance between tiles i and j. This 
matrix is incorporated into the eigendecomposition step. The wSIR output has the 
same structure as the SIR output. 

## Method demonstration

We will begin the workshop by going through some steps in the wSIR algorithm. You do not need to understand the steps in the method to be able to use the package - we only include this to give you an understanding of how it works 

### Tile creation

The key conceptual difference between SIR and wSIR is in how we estimate E(X|Y). 

In both cases, this begins by grouping the cells into tiles according to their spatial positions. We split the data into rectangular tiles, with an equal number of "cuts" in both spatial axes.

We will work here on the spatial coordinates of embryo 3, as this is the sample with the most cells.

```{r}
ggplot(coords3, aes(x = x, y = y)) +
  geom_point(size = 0.5) +
  theme_classic() + 
  labs(x = "X spatial coordinate", y = "Y spatial coordinate") +
  ggtitle("Embryo 3 spatial coordinates")
```

We will then split the data into 9 tiles. 

This means assigning each cell to one tile of $(1,1), (1,2), (1,3), (2,1), (2,2), (2,3), (3,1), (3,2), (3,3)$, where $(1,1)$ is the tile in the top left, $(1,3)$ is in the top right, and so on. We do this by assigning all cells in the upper third of the sample (according to spatial position) to one of $(1,1), (1,2), (1,3)$, then the middle third to one of $(2,1), (2,2), (2,3)$, and the lower third to one of $(3,1), (3,2), (3,3)$. This is done so that each of those thirds is as equally populated as possible. Next, a similar procedure is applied to the left third, the middle third, and the right third. 

::: question
**Questions**

1. Is this procedure ensuring an equal number of cells in all tiles?
2. Try modifying the chunk below to allocate the cells to tiles in different ways.
:::

Note the code to do this in the wSIR package has been optimised - the code here is only to demonstrate the concept. 

```{r}
# Define the number of cells in each third of the sample

third1 <- floor(n3/3)
third2 <- floor(n3/3)
third3 <- n3 - third1 - third2

# create new dataframe with row IDs, which we sort by later 
with_id <- data.frame(x = coords3$x,
                      y = coords3$y,
                      id = c(1:n3))

# sort by x spatial coordinate to find the tile coordinates in the x spatial axis
x_sorted <- arrange(with_id, x)
x_sorted$x_tile <- c(rep(1, third1),
                     rep(2, third2),
                     rep(3, third3))

# sort by y spatial coordinate to find the tile coordinates in the y spatial axis
y_sorted <- arrange(x_sorted, y)
y_sorted$y_tile <- c(rep(1, third1),
                     rep(2, third2),
                     rep(3, third3))

# sort data back into original order so the tile_allocations vector is in the same order as the expression and coordinates data
id_sorted <- arrange(y_sorted, id)
tile_allocations <- paste0(id_sorted$x_tile, ", ", id_sorted$y_tile)

# Plot cells in spatial positions coloured by tile allocation
ggplot(data.frame(x = coords3$x, 
                y = coords3$y, 
                tile = tile_allocations), aes(x = x, y = y, colour = tile)) +
  geom_point() +
  labs(x = "X spatial axis", y = "Y spatial axis", colour = "Tile alocation") +
  theme_classic() +
  ggtitle("Cells from embryo 3 coloured by tile allocation")
```

### Weight matrix creation

Once the cells are allocated to tiles, the weight matrix is then created. For this workshop, we will use the `CreateWeightMatrix` function from the wSIR package. Note this is a hidden function as users do not need to use it, but we will access it anyway for this workshop.

You can adjust the alpha parameter, which represents the strength of spatial weighting, and then visualise the resulting weight matrix. 

::: question
**Questions**

1. Observe what happens when you have 0 spatial weighting, or very strong spatial weighting.
2. If you had never heard of wSIR and were asked to quantify the spatial proximity of the tiles above in a matrix with values between 0 and 1, how would it look? Is there a value of alpha that would produce such a matrix?
:::

If you want to adjust the tile allocations, it is best to do that in the previous chunk. Then you can simply run this next chunk to view the resulting weight matrix. 

```{r}
# Modify tile allocations to suit the required input for CreateWeightMatrix function.
# As with other components in this section, you won't need to know this to use the wSIR package - it is just included in case you are very curious about how the functions work. 

# requires dataframe input with a column named "coordinate". Tile allocations must have 3 "axes": one for each spatial axis, and a third for the sample each cell belongs to. In this case, all cells are from the same sample, which is why I specify 1 in the next line. This would change if we used multiple samples at once, in which case we need to set the weight value between tiles of differing samples to 0.

tile_allocations_df <- data.frame(coordinate = paste0(tile_allocations, ", 1"))

weight_matrix <- wSIR:::createWeightMatrix(coords = coords3, 
                                           labels = tile_allocations_df, 
                                           alpha = 4) ## Adjust this parameter!

n_tiles <- nrow(weight_matrix)
# Turn weight matrix into a format that can be visualised in a ggplot heatmap

w_mat_df <- data.frame(value = as.vector(weight_matrix),
                       x = vec_rep_each(c(1:n_tiles), n_tiles),
                       y = rep(c(1:n_tiles), n_tiles))
ggplot(w_mat_df, aes(x = as.factor(x), y = as.factor(y), fill = value)) +
  geom_tile() +
  scale_fill_viridis_c() +
  theme_classic() +
  labs(x = "Tile", y = "Tile", fill = "Weight matrix value") +
  ggtitle("Example weight matrix")

```


The next steps in the wSIR method include more complicated linear algebra, which we will not cover in this workshop.

This is the end of the method demonstration section. Up to now, we hope you have an understanding of how the tiles are allocated, and how the weight matrix will look with varying allocations and spatial weighting strengths. 

## wSIR parameter exploration

Here, we demonstrate the wSIR function `exploreWSIRParams`, which shows the impact of changing the parameter values on the distance correlation between the wSIR low-dimensional embedding and the true spatial coordinates. A higher distance correlation corresponds to a better performance. 

Distance correlation is a measure of the correlation between two distance matrices. Ideally, if two cells are spatially close, we would want them to be close in the low-dimensional embedding, and similar for two distant cells - distance correlation is a measure of this. If you'd like more details on this metric, there are many good online resources. 

The following chunk has some parameters we you can explore. You can vary the alpha and slice values that are tuned over, to see the performance with different combinations. This can take a while to run, so we suggest you randomly subset some of the data with the variable `subsetted`. In practice, this doesn't change the results of your study, we only do it as computing the distance correlation can be quite slow over many cells.  

We recommend you don't adjust `nrep` or `varThreshold`, as this can make it take a very long time to run. Be aware that the following chunk can take some time to run, so be careful with what you change!

```{r}
subsetted = 0.2 # Change this to specify the proportion of the data you want to use for this exploration
rsample <- sample(c(TRUE, FALSE), size = n3, replace = TRUE, prob = c(subsetted, 1-subsetted))
# FIXME
EWP_object <- exploreWSIRParams(exprs = exprs3[rsample,],
                                coords = coords3[rsample,],
                                nrep = 3, # This function computes a random train/test split of the data nrep times
                                varThreshold = 0.95,
                                alpha_vals = c(0,4,10,20),
                                slice_vals = c(5,10,15,20))

ggplot(EWP_object$results_dataframe, aes(x = alpha, y = slices, fill = value)) +
  geom_tile() +
  theme_classic() +
  scale_fill_viridis_c() +
  labs(x = "Alpha", y = "Slices", fill = "Distance Correlation")
```

The key takeaway you will (hopefully) see above is that wSIR has relatively invariant distance correlation for moderate values of slices and alpha. This is in contrast to the column representing `alpha = 0`, which means there is no spatial weighting at all (recall the corresponding weight matrix). Why do you think performance deteriorates with an increasing number of slices when `alpha = 0`?

## wSIR algorithm

Using the wSIR algorithm itself is quite straightforward. For users, there is a single function to compute the low-dimensional embedding and the loadings matrix, which is named wSIR. This function only needs the gene expression matrix and the spatial coordinates dataframe as input, though you can specify parameters to the non-default values if desired. The wSIR function returns a list of results with 5 (named) slots, whose details can be found at `?wSIR::wSIR`.

```{r}
# wSIR with default parameter values
wsir_object <- wSIR::wSIR(X = exprs3, 
                          coords = coords3)

# wSIR selecting to optimise over a range of parameters using exploreWSIRParams
## Not recommended in practice since wSIR performance is mostly consistent when you select reasonable values for slices and alpha
wsir_object <- wSIR::wSIR(X = exprs3, 
                          coords = coords3,
                          optim_params = TRUE,
                          nrep = 3,
                          alpha_vals = c(0,5,10),
                          slice_vals = c(8,12,15),
                          varThreshold = 0.95)

# wSIR specifying some parameter values
wsir_obj <- wSIR::wSIR(X = exprs3, 
                       coords = coords3,
                       alpha = 5,
                       slices = 10,
                       varThreshold = 0.95)
```

## wSIR application: interpretability

The wSIR package includes some functions to give an insight into what biological information the method is using. These functions are firstly for interpretability, so you can understand which genes are the most important for the low-dimensional space. Secondly, these functions could be used to give more biological understanding, as you can find how genes impact and are impacted by the spatially-aware low-dimensional embedding. 

### wSIR Top Genes

The `findTopGenes` function finds and plots the genes with highest loading in 
the specified wSIR directions (default is direction 1). If a gene has high 
loading (in terms of magnitude), it is more important to the wSIR direction. 
Since the wSIR directions are designed to retain information about each cell's 
spatial position, the genes with high loading should be spatially-related genes. 

In the plot below, we can see which genes have the highest loading in wSIR 
direction 1. This is useful as it gives us an intuition about how wSIR creates 
the low-dimensional embedding. We can see that some of the genes are known 
spatial genes (e.g. Cdx2, Hox-), which is what we would expect to see.

```{r}
top_genes_obj = findTopGenes(WSIR = wsir_obj, highest = 8) # create top genes object
top_genes_plot = top_genes_obj$plot # select plot
top_genes_plot # print plot

top_genes_obj = findTopGenes(WSIR = wsir_obj, highest = 8, dirs = 2:4)
top_genes_plot = top_genes_obj$plot
top_genes_plot
```

### Visualising wSIR Scores 

The `visualiseWSIRDirections` function plots each cell at its spatial position, 
coloured by its value for each of the specified wSIR columns. This gives us an 
understanding of what each column of the low-dimensional embedding represents. 

Below, we visualise the cells at their spatial positions, coloured by each of 
the 5 wSIR directions The top left plot illustrates how, for this example, 
wSIR direction 1 captures information about the "y" spatial axis, since cells 
with higher "y" coordinate have low wSIR1 value, while cells with lower "y" 
coordinate have higher wSIR1 value. wSIR2 is shown in the next plot over 
(the one titled "2"), and we can see that wSIR column two appears to capture 
information about the "x" spatial coordinate. The remaining three wSIR columns 
all contain information about cell types, which we can tell by the regions of 
high and low wSIR column values spread across the tissue. 

```{r}
vis_obj = visualiseWSIRDirections(coords = sample1_coords, WSIR = wsir_obj, dirs = 8) # create visualisations
vis_obj
```

### UMAP on low-dimensional embedding

The two functions `generateUmapFromWSIR` and `plotUmapFromWSIR` create and 
display UMAP dimension reduction calculated on the wSIR low-dimensional 
embedding. We can colour the UMAP plot (where each point represents a cell) by 
its value for various genes of interest. This visualises the structure of the 
wSIR dimension reduction space, which is useful to gain more intuition about 
what the space represents. Specifically, we can see if the wSIR space contains 
neighbourhoods of high expression for specific genes, thus better understanding 
how this space is made. 

To specify which genes we would like to include, we can use the output from 
the `findTopGenes` function from above, which finds spatially-related genes by 
ranking those with the highest loading in relevant wSIR directions. This output 
is then the value for the `highest_genes` parameter. Otherwise, we could also 
specify our own genes of interest if there are some specific genes we would 
like to visualise. For example, if we wanted to visualise the expression 
distribution for Cdx2 and Hoxb4, we could use `genes = c("Cdx2", "Hoxb4")` as 
an argument in `plotUmapFromWSIR` (and leave `highest_genes` blank). 

Below, we use the UMAP function to visualise the wSIR space computed on the 
gene expression data from sample 1. We colour each cell by their values for 
the 6 genes with highest value in wSIR direction 1 (as found by the 
`findTopGenes` function previously). We can see that for some of these genes, 
there are specific regions of high expression in the UMAP plots, suggesting 
that the wSIR space separates cells based on their expression for those genes.

```{r}
umap_coords = generateUmapFromWSIR(WSIR = wsir_obj)
umap_plots = plotUmapFromWSIR(X = sample1_exprs,
                              umap_coords = umap_coords,
                              highest_genes = top_genes_obj,
                              n_genes = 6)
umap_plots
```

## wSIR application: improving performance

Aside from improving interpretability and aiding biological insight, wSIR can improve modelling performance. For users, you may want to project single-cell gene expression data into a spatially-aware low-dimensional embedding, to which you can apply any machine learning methods (e.g cell-type classification, location prediction, etc). 

Here, we first demonstrate how to project new single-cell data into a spatially aware low-dimensional space, using out wSIR object from a previous section.

We will use the gene expression matrix from mouse embryo 1 here.

```{r}
dim(exprs1)
dim(exprs3) # Note that the genes in your spatial and single-cell samples must be the same, but there can be different numbers of cells.

exprs1_projected <- projectWSIR(wsir = wsir_obj, newdata = exprs1)

dim(exprs1_projected)
```

From just that line, you can now apply downstream analysis to this low-dimensional embedding of embryo 1's gene expression data.

Here, we will show how this method can be applied to Tangram, a popular deep-learning-based spatial alignment method. Tangram is available here: https://www.nature.com/articles/s41592-021-01264-7 .

For this simulation, we will use embryo 3 as our spatial sample, and embryo 1 as our single-cell sample. The method workflow is detailed below. The concept will be to use the wSIR low-dimensional embedding to predict the spatial coordinates of the single-cell data.

![Tangram downstream analysis example workflow](/tangram_workflow.png)

Note that for this workshop, we will not actually compute the Tangram predicted coordinates, as it takes too long to run. We will load in the predicted coordinates which we have already computed, and we will create a results plot from that. 

Below loads in 7 matrices, all of dimension n1 by 2, containing the predicted coordinates using as inputs: PCA, PLS, SIR, wSIR, LDA, counts and logcounts. The file names are of the form `pred_pca_em1`, in that case for the predicted coordinates of embryo 1 using the PCA low-dimensional embedding as the Tangram input. We also include the predicted coordinates using just counts or LogCounts as the inputs (without any dimension reduction applied) as those are the default inputs for Tangram.

```{r}
#data(em1_tangram_preds_red) # This loads a list (not vector) of predicted coordinates into your environment, named pred_em1_tangram_red
load(system.file("extdata", "em1_tangram_preds_red.RData", package="BioCAsia2024wSIR"))
```

To evaluate, we will compute the distance correlation between the predicted and the actual coordinates, for the predicted coordinates from all dimension reduction methods. This is not part of the wSIR package, but should demonstrate the effectiveness of using wSIR as a dimension reduction tool to improve downstream analysis.

```{r}
dim(wsir_obj$scores) # This is the features matrix for the training set
dim(coords3) # This is the spatial coordinates for the training set
dim(exprs1_projected) # This is the features matrix for the testing set, with which we will use Tangram to predict its spatial coordinates

# save predicted coordinates in a list for easy results generation


em1_dcs <- rep(0,7) # initialise empty vector to fill with the distance correlations
i = 0
for (pred in pred_em1_tangram_red) {
  print(i)
  i = i + 1
  em1_dcs[i] <- Rfast::dcor(pred, coords1)$dcor
}

tangram_distcor_df <- data.frame(distcor = em1_dcs,
                                 tangram_input = c("LDA", 
                                                   "wSIR", 
                                                   "SIR",
                                                   "PCA",
                                                   "PLS",
                                                   "Counts",
                                                   "LogCounts"))
ggplot(tangram_distcor_df, aes(x = tangram_input, y = distcor)) +
  geom_col(width = 0.5) +
  theme_classic() +
  labs(x = "Tangram Input", y = "Distance Corrrelation") +
  ggtitle("Distance correlation between predicted and actual coordinates")
```

### wSIR application: multiple samples

In the previous section, we used the gene expression data and the spatial coordinates of one spatial sample to project the gene expression data of a different sample into low-dimensional space. However, we may often have access to multiple spatial samples. wSIR is able to incorporate multiple such samples, by setting the weight matrix value for two tiles from different spatial samples to be 0. The code to use, for example, the gene expression and spatial coordinates for embryos 2 and 3 to project the gene expression data of embryo 1 into low-dimensional space is below.

```{r}
wsir_obj_2_3 <- wSIR::wSIR(X = rbind(exprs2, exprs3),
                           coords = rbind(coords2, coords3),
                           samples = c(rep(2, n2), rep(3, n3)), # specify samples
                           alpha = 4,
                           slices = 10,
                           varThreshold = 0.95)

proj1_from23 <- projectWSIR(wsir = wsir_obj_2_3, newdata = exprs1)
dim(proj1_from23) # gene expression data from sample 1 projected into low-dimensional space
```

We will not explore it here, but our results (manuscript on the way!) show that wSIR performance improves when you incorporate multiple spatial samples. Therefore, if you are in a situation where you do have multiple spatial samples, it is best to use wSIR and account for the multiple samples as above. 








